P5-jdk-version-support.md

```diff
@@ -0,0 +1,126 @@
+Java version becomes a major undertaking. For library authors it was hard to
+know when it was "safe" to drop support for an old JDK version, and may find out
+the issues caused years after the decision. This created an ecosystem of sincere
+fear, uncertainty, and doubt where library owners did not want their
+dependencies to drop older JDK support, even though the benefit was unclear. It
+was rightly known that _some_ user was certainly still using an older JDK
+version, but it was unclear how many and whether they ever would need a new
+version of a library.
+
+That process has been changing since Java 8, especially with the new more rapid
+releases and planned LTS releases and reduced support timelines. Java releases
+have been offering features that have been encouraging upgrades and Java is
+hiding more of its internals to make upgrades easier. Running separate
+applications in a single, large application server is less common which allows
+upgrading pieces individually. The ecosystem has gotten more practiced at
+supporting new versions, less willing to be held back, and better at
+compartmentalizing legacy applications.
+
+gRPC is widely used and is lower in the dependency tree, so it has more
+conservative support. gRPC Java 1.0 supported Java 6, which was dropped
+following discussion in grpc/grpc-java#3961 because Java 6 was becoming
+unsupportable as the compiler was unmaintained and other libraries were dropping
+support. The discussion for dropping Java 7 is happening at grpc/grpc-java#4671
+and is necessary for reasons similar to Java 6. However, both processes were
+time-consuming and had no predictable timeline.
+
+gRPC has suffered from supporting old Java versions. There are hidden
+maintenance costs, but it is easy for someone to argue those should just be paid
+(either because of gRPC's scale or because they are not paying those costs).
+However, the public API surface has been limited which directly impacts users.
+The most obvious example of this is the lack of generated `interface`s for
+servers. Adding methods to an existing service must be backward-compatible, and
+the only way to do that before Java 8 is with an abstract class. But once we can
+use Java 8 language features we can use `default` methods in an `interface`.
+
+Android support is tracked separately and generally follows Google Play
+Services's supported API level, for security and as a reasonable guideline of
+what is worth supporting. Android support is not impacted by this gRFC as it has
+different options available to it (e.g., desugaring) and does not align cleanly
+to JDK versions.
+
+## Proposal
+
+gRPC Java may drop support for a JDK version when that version is no longer
+receiving [Premier Support from Oracle][Oracle Java support]. When doing so,
+gRPC maintainers will publicize a recent minor version release (but not
+necessarily the most recent) to be the suggested release for users on the older
```

Comment:
What's the concern if we _always_ choose the most recent minor version to publicize? Doing so we will never require users to downgrade. (For the upcoming release, even though we claim not to support the old JDK, we don't have to immediately migrate to build it with the new JDK or to use the new language features.)

Reply:
I think we will prefer a very recent version, but I think we'd also take into account something that seems to have been heavily used (and thus, we are more confident is stable). So we may use a slightly older version if it happened to have been heavily used with Google Client Libraries, Beam, or similar. But we'd still prefer the newest version that seemed to have been heavily used.

The downgrade problem isn't expected to be too severe for shorter periods because we've seen devs in these circumstances tend not to update frequently (which is also why it seems they don't really need features).